### ISSUE TITLE
Consider adding Kill Workflow action to Workflow menu

### ISSUE THREAD

**Issue by jonnew** [bonsai-rx/bonsai#1868]
There are situations where workflows can enter an infinite loop. This is usually due to buggy third party code. It can occur when there there are deadlocks, e.g. when performing detailed hardware interaction that requires tricky interaction with third-party libraries that are not thread safe. Of course, there are many other examples of how this might happen.

As far as I understand, the Bonsai visual editor is an IDE for the the bonsai language. This IDE can be used to edit, compile, and run code in the same way that e.g. Visual Studio can do the same for written languages. Unlike most IDEs that I have encountered, the Bonsai IDE does not seem to have a way to kill a misbehaving process without closing the editor. By analogy, this would be the same has needing to close Visual Studio in order to stop deadlocked process it has spawned. In this situation, the Stop button becomes unresponsive to user input. The user must eventually come to terms with the fact that something has silently gone wrong and kill the process by clicking 'X' on the console window in the background. In my opinion:

- To an average user, this feels like Bonsai itself is misbehaving, leading to (unwarranted) bad feelings about its reliability
- The process for escaping the situation itself feels draconian and is potentially going to result in the loss of unsaved edits to their workflow without any ability to escape the situation
- The process is far from obvious

The correct solution in this situation, of course, is to fix the underlaying issue. I feel that a properly implemented Kill Workflow action could both aide in achieving this, while providing significant quality of life improvement for Bonsai users. To do this, this action can have the following properties

1. It can be "hidden" as a dropdown item in the Workflow menu in order to discourage its regular use
2. It can spawn a warning dialog that informs the user they are about to take an extreme action.
3. It can save the misbehaving workflow and, potentially information about its run state, in a log file that can be used to find the underlying issue.

**Comment by PathogenDavid** [comment]
We talked about this a bit during DCM today.

This feature is actually somewhat difficult to provide. When you hit the kill button in most IDEs, you're killing the process being debugged. In Bonsai, your workflow runs within your editor process. When things are in a state where the stop button doesn't work, there's not actually much the editor can do. Additionally, depending on the fault the UI might not actually be usable because the Windows message pump is starved.

(If you've used Unity before, it has this issue as well. Bugs in your game can take down the editor or cause it to lock up.)

Potentially strategies discussed:
* Run the workflow out of process, handle UI via shared memory as I did in [this experiment](https://github.com/PathogenDavid/MultiProcessImGui) (which I made basically for this exact scenario.)
    * Pros:
        * Very flexible while still maintaining process boundaries.
        * Less limiting in the future.
        * Not too complicated (but maybe only if you're David.)
    * Cons:
        * Certain fatal UI mistakes in the child process can still take out the parent process.
        * Very heavily tied to a specific flavor of UI framework (like Dear ImGui.)
        * Slower workflow start-up time (have to reload all package assemblies and re-JIT them.)
* Run the workflow out of process, handle UI via serialized messages back and forth. Visualizers still run in-process but must be separate windows.
    * Pros:
        * Less complex to reason about, closer to how most people solve problems like this.
        * Process boundaries are very strongly defined, if the child process can take out the parent process it's due to a bug in the parent process.
    * Cons:
        * Visualizers cannot participate in main window interactions. (They can't be docked, they can't be embedded in the workflow - something I've pondered about.)
        * More tedious to set up and extend. Maybe fine if we only consider the property panel.
        * Most likely not friendly towards [the workflow state watch](https://github.com/bonsai-rx/bonsai/pull/1975) (Right now rendering is the bottleneck but this would make IPC the bottleneck.)
        * Slower workflow start-up time (have to reload all package assemblies and re-JIT them.)
* Killing the workflow actually also the editor and then automatically reopens it to the workflow you were working on.
    * Requires some variation of https://github.com/bonsai-rx/bonsai/issues/2077
    * Pros:
        * Simplest to reason about
        * Fits well into our current multi-process model (otherwise this would be a con due to the lack of a supervisor process)
    * Cons:
        * More strongly ties us to the model of having the editor be multi-process (the reason we did this originally is not 100% valid)
        * If the UI is dead, you can't interact with the button to kill the workflow (see below for potential workarounds)
        * Not the best user experience since at first glance it might seem like the editor crashed when you killed the workflow
* Add an option to Bonsai to default to opening the editor to the last known workflow (and users just kill Bonsai and reopen it)
    * Already possible today, but lessens friction somewhat
    * Even if we don't make this a setting, could offer to do it when Bonsai did not close cleanly last time it was used. (Lots of apps do this)
    * Pros:
        * Very simple, useful for other reasons
    * Conss:
        * Forces a particular way of working on users
        * Not totally seamless

Possible workarounds for when the UI is unresponsive:
* One possible workaround might be to allow a keyboard shortcut known by the bootstrapper
    * In dev club I described this as a keyboard shortcut on the terminal, but we could theoretically have the bootstrapper listen for keyboard shortcuts on Windows.
        * Handling this sort of hotkey properly is weird though. (`RegisterHotKey` requires a single process handle the hotkey, which might mean an additional supervisor process shared between Bonsai instances. Raw input is doable but somewhat cumbersome, I have a personal C# project that does hotkeys this way though.)
    * A keyboard shortcut on the terminal is doable, but is very hard to discover and unintuitive. I'd also personally like to see the terminal go away.
* Allow killing workflows from a notification tray icon
    * By default results in tons of Bonsais in the notification tray (annoying for power users, need to make sure they can be disambiguated)
    * Could have a supervisor process submit the tray icon and the context menu lists all active Bonsai editors/workflows. (More complex to implement.)
    * Somewhat Windows-centric

---------------------------------

Another idea I had which was not discussed during dev club:
* Have the bootstrapper process act as a watchdog, if the Bonsai instance was last known to be running a workflow and has not sent a heartbeat in too long. Automatically present a dialog suggesting Bonsai is stuck and offer to restart it. (Similar to the Windows "The process is not responding" dialog, or [this custom one from VSCode](https://github.com/user-attachments/assets/05789733-0123-4996-8daa-63480d962dd3).)
    * Pros:
        * Pretty simple to implement, doesn't suffer from discoverability issues
    * Cons:
        * False positives possible. (Especially if the timeout is too short, but if it's too long it can be frustrating having a stuck Bonsai for 30 seconds waiting for the dialog to appear.)
        * Theoretically possible to have an unresponsive UI but still have a living message pump (TBH not sure what's going on when this happens, just that I've seen it before in other apps)

---------------------------------

My personal vibe is that while having the workflow run out of process is desirable in some ways, just reopening the editor is probably the simpler solution.

**Comment by jonnew** [comment]
Thank you very much @PathogenDavid for the clear explanation. This explains a lot and makes me much more sympathetic to the situation since this is obviously hard to do correctly.

I think the last sentence, unfortunately, is probably correct. The correct solutions seem like a lot of work and with introduce other issues. I think perhaps this might be OK though:

> Have the bootstrapper process act as a watchdog, if the Bonsai instance was last known to be running a workflow and has not sent a heartbeat in too long. Automatically present a dialog suggesting Bonsai is stuck and offer to restart it. (Similar to the Windows "The process is not responding" dialog, or [this custom one from VSCode](https://github.com/user-attachments/assets/05789733-0123-4996-8daa-63480d962dd3).)

However, since receiving that notification is almost more annoying than the situation as it is now, I might change the dialog presentation a bit: silently detect the loss of heartbeat and let the user close the workflow as they need to now. However, the next time the bonsai editor is opened, present a dialog indicating:

> "It seems like bonsai may have stopped responding last time, do you want to load the this autosaved workflow"

or something. That's the way KiCAD does it when it crashes which happens maybe 10 times a day for release candidates.

**Comment by jonnew** [comment]
yes that makes sense. This is how KiCAD does it. One side effect that is might be worth mentioning:

1.  The user starts their workflow
2. While its running they copy the directory that holds it into a thumbdrive or perform some periodic backup of the directory to a server if datafiles are being generated during an automated experiment.
3. They or someone else downloads the directory later and opens the workflow.
4. They are presented with the dialog since the sentinel file was copied which is a bizarre user experience. 

Its probably fine, but it happened to me enough with  KiCAD and was confusing enough I spent time figuring out I was copying a `kicad.lck` file with my designs because KiCAD was open while copy was performed.
